File: mobile/ios/Runner/SpeechSynthesizer.swift
import AVFoundation
import Flutter

/// Reads a finalized translation aloud with the system voice.
///
/// Deliberately does NOT touch the AVAudioSession: while a translation session
/// is running, AudioCaptureManager owns the session (.playAndRecord with
/// .measurement, .defaultToSpeaker) and reconfiguring it here would disturb
/// live capture. AVSpeechSynthesizer plays into whatever session is already
/// active, which is exactly what we want.
///
/// Emits {"speaking": true/false} so Dart can quiet the microphone uplink for
/// the duration and reopen it the instant speech ends.
final class SpeechSynthesizer: NSObject, FlutterStreamHandler,
  AVSpeechSynthesizerDelegate
{
  private let synthesizer = AVSpeechSynthesizer()
  private var eventSink: FlutterEventSink?
  private var preparedVoice: AVSpeechSynthesisVoice?

  override init() {
    super.init()
    synthesizer.delegate = self
  }

  // MARK: - FlutterStreamHandler

  func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink)
    -> FlutterError?
  {
    eventSink = events
    return nil
  }

  func onCancel(withArguments arguments: Any?) -> FlutterError? {
    eventSink = nil
    return nil
  }

  // MARK: - Speaking

  /// Resolves (and caches) the voice for `languageCode` ahead of the first
  /// tap. AVSpeechSynthesizer itself needs no warming, but looking the voice
  /// up here keeps the first speak() off the voice-catalog path.
  func prepare(languageCode: String) {
    preparedVoice = Self.voice(for: languageCode)
  }

  /// Speaks `text` in `languageCode` (BCP-47). Returns false when the device
  /// has no voice for that language, so the UI can say so instead of appearing
  /// to do nothing.
  func speak(text: String, languageCode: String) -> Bool {
    let trimmed = text.trimmingCharacters(in: .whitespacesAndNewlines)
    guard !trimmed.isEmpty else { return false }
    // Use the warmed voice when it matches, otherwise resolve now.
    let voice: AVSpeechSynthesisVoice
    if let prepared = preparedVoice,
      prepared.language.caseInsensitiveCompare(languageCode) == .orderedSame
    {
      voice = prepared
    } else if let resolved = Self.voice(for: languageCode) {
      voice = resolved
      preparedVoice = resolved
    } else {
      return false
    }

    // A new tap replaces whatever is being said rather than queueing behind it.
    if synthesizer.isSpeaking {
      synthesizer.stopSpeaking(at: .immediate)
    }
    let utterance = AVSpeechUtterance(string: trimmed)
    utterance.voice = voice
    utterance.rate = AVSpeechUtteranceDefaultSpeechRate
    synthesizer.speak(utterance)
    return true
  }

  func stop() {
    guard synthesizer.isSpeaking else { return }
    synthesizer.stopSpeaking(at: .immediate)
  }

  /// Exact match first ("pt-BR"), then any installed voice for the primary
  /// subtag ("pt"), so a catalog code still speaks on a device that only has a
  /// regional variant installed.
  private static func voice(for languageCode: String) -> AVSpeechSynthesisVoice? {
    if let exact = AVSpeechSynthesisVoice(language: languageCode) {
      return exact
    }
    let primary = languageCode.split(separator: "-").first.map(String.init)
      ?? languageCode
    if let match = AVSpeechSynthesisVoice.speechVoices().first(where: {
      $0.language.lowercased().hasPrefix(primary.lowercased())
    }) {
      return match
    }
    return AVSpeechSynthesisVoice(language: primary)
  }

  // MARK: - AVSpeechSynthesizerDelegate

  func speechSynthesizer(
    _ synthesizer: AVSpeechSynthesizer, didStart utterance: AVSpeechUtterance
  ) {
    send(speaking: true)
  }

  func speechSynthesizer(
    _ synthesizer: AVSpeechSynthesizer, didFinish utterance: AVSpeechUtterance
  ) {
    send(speaking: false)
  }

  func speechSynthesizer(
    _ synthesizer: AVSpeechSynthesizer, didCancel utterance: AVSpeechUtterance
  ) {
    send(speaking: false)
  }

  private func send(speaking: Bool) {
    guard let sink = eventSink else { return }
    DispatchQueue.main.async { sink(["speaking": speaking]) }
  }
}

File: mobile/android/app/src/main/kotlin/com/livetranslator/live_translator/SpeechSynthesizer.kt
package com.livetranslator.live_translator

import android.content.Context
import android.os.Handler
import android.os.Looper
import android.speech.tts.TextToSpeech
import android.speech.tts.UtteranceProgressListener
import io.flutter.plugin.common.EventChannel
import java.util.Locale

/**
 * Reads a finalized translation aloud with the device's own voice
 * (android.speech.tts.TextToSpeech).
 *
 * Emits {"speaking": true/false} so Dart can quiet the microphone uplink while
 * the phone is talking and reopen it the moment speech ends. Nothing here
 * touches capture or the Gemini socket: speaking never interrupts a listening
 * session.
 */
object SpeechSynthesizer : EventChannel.StreamHandler {

    private const val UTTERANCE_ID = "sayvo.translation"

    private var tts: TextToSpeech? = null
    private var ready = false
    private var eventSink: EventChannel.EventSink? = null
    private val main = Handler(Looper.getMainLooper())

    /** Speak request that arrived before the engine finished initializing. */
    private var pending: Pair<String, String>? = null

    fun init(context: Context) {
        if (tts != null) return
        tts = TextToSpeech(context.applicationContext) { status ->
            ready = status == TextToSpeech.SUCCESS
            if (!ready) return@TextToSpeech
            tts?.setOnUtteranceProgressListener(object : UtteranceProgressListener() {
                override fun onStart(utteranceId: String?) = emit(true)
                override fun onDone(utteranceId: String?) = emit(false)
                override fun onStop(utteranceId: String?, interrupted: Boolean) = emit(false)

                @Deprecated("Required by the base class", ReplaceWith(""))
                override fun onError(utteranceId: String?) = emit(false)
                override fun onError(utteranceId: String?, errorCode: Int) = emit(false)
            })
            // A tap that landed during initialization still gets spoken.
            pending?.let { (text, language) ->
                pending = null
                speak(text, language)
            }
        }
    }

    /**
     * Warms the engine and pre-selects the voice for [languageCode]. Android's
     * TextToSpeech binds to its engine service asynchronously, so doing this
     * when a session starts is what keeps the FIRST speaker tap instant.
     */
    fun prepare(languageCode: String) {
        val engine = tts ?: return
        if (!ready) return
        applyLanguage(engine, languageCode)
    }

    /**
     * Speaks [text] with the voice for [languageCode] (BCP-47, e.g. "ar-SA",
     * "pt-BR"). Returns false when no voice is installed for that language, so
     * the UI can say so rather than appearing to do nothing.
     */
    fun speak(text: String, languageCode: String): Boolean {
        val engine = tts ?: return false
        val trimmed = text.trim()
        if (trimmed.isEmpty()) return false
        if (!ready) {
            // Remember it; the init callback above will speak it.
            pending = trimmed to languageCode
            return true
        }
        if (!applyLanguage(engine, languageCode)) return false
        // QUEUE_FLUSH: a new tap replaces what is being said, never queues.
        val result = engine.speak(trimmed, TextToSpeech.QUEUE_FLUSH, null, UTTERANCE_ID)
        return result == TextToSpeech.SUCCESS
    }

    fun stop() {
        tts?.stop()
        emit(false)
    }

    /** Exact locale first ("pt-BR"), then the plain language ("pt"). */
    private fun applyLanguage(engine: TextToSpeech, languageCode: String): Boolean {
        val exact = Locale.forLanguageTag(languageCode)
        if (engine.isLanguageAvailable(exact) >= TextToSpeech.LANG_AVAILABLE) {
            engine.language = exact
            return true
        }
        val primary = Locale.forLanguageTag(languageCode.substringBefore('-'))
        if (engine.isLanguageAvailable(primary) >= TextToSpeech.LANG_AVAILABLE) {
            engine.language = primary
            return true
        }
        return false
    }

    private fun emit(speaking: Boolean) {
        main.post { eventSink?.success(mapOf("speaking" to speaking)) }
    }

    override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
        eventSink = events
    }

    override fun onCancel(arguments: Any?) {
        eventSink = null
    }
}
